presto-main/src/main/java/io/prestosql/operator/GroupedTopNRowNumberAccumulator.java

```diff
@@ -0,0 +1,483 @@
+     * the root position, and applying a heap down bubbling operation to heap-ify.
+     *
+     * @param groupId
+     * @param newRowId
+     * @param contextEvictionListener optional callback for the root node that gets popped off
+     */
+    private void heapPopAndInsert(long groupId, long newRowId, @Nullable LongConsumer contextEvictionListener)
+    {
+        long heapRootNodeIndex = groupIdToHeapBuffer.getHeapRootNodeIndex(groupId);
+        checkState(heapRootNodeIndex != UNKNOWN_INDEX, "popAndInsert() requires at least a root node");
+
+        // Clear contents of the root node to create a vacancy for another row
+        long poppedRowId = heapNodeBuffer.getRowId(heapRootNodeIndex);
+
+        long currentNodeIndex = heapRootNodeIndex;
+        while (true) {
+            long maxChildNodeIndex = heapNodeBuffer.getLeftChildHeapIndex(currentNodeIndex);
+            if (maxChildNodeIndex == UNKNOWN_INDEX) {
+                // Left is always inserted before right, so a missing left child means there can't be a right child,
+                // which means this must already be a leaf position.
+                break;
+            }
+            long maxChildRowId = heapNodeBuffer.getRowId(maxChildNodeIndex);
+
+            long rightChildNodeIndex = heapNodeBuffer.getRightChildHeapIndex(currentNodeIndex);
+            if (rightChildNodeIndex != UNKNOWN_INDEX) {
+                long rightRowId = heapNodeBuffer.getRowId(rightChildNodeIndex);
+                if (rowComparator.compare(rightRowId, maxChildRowId) > 0) {
+                    maxChildNodeIndex = rightChildNodeIndex;
+                    maxChildRowId = rightRowId;
+                }
+            }
+
+            if (rowComparator.compare(newRowId, maxChildRowId) >= 0) {
+                // New row is greater than or equal to both children, so the heap invariant is satisfied by inserting the
+                // new row at this position
+                break;
+            }
+
+            // Swap the max child row value into the current node
+            heapNodeBuffer.setRowId(currentNodeIndex, maxChildRowId);
+
+            // Max child now has an unfilled vacancy, so continue processing with that as the current node
+            currentNodeIndex = maxChildNodeIndex;
+        }
+
+        heapNodeBuffer.setRowId(currentNodeIndex, newRowId);
+
+        if (contextEvictionListener != null) {
+            contextEvictionListener.accept(poppedRowId);
+        }
+    }
+
+    /**
+     * Buffer abstracting a mapping from group ID to a heap. The group ID provides the index for all operations.
+     */
+    private static class GroupIdToHeapBuffer
+    {
+        private static final long INSTANCE_SIZE = ClassLayout.parseClass(GroupIdToHeapBuffer.class).instanceSize();
+
+        /*
+         *  Memory layout:
+         *  [LONG] heapNodeIndex G1,
+         *  [LONG] heapNodeIndex G2,
```

Comment:
I'm not sure what this means

Reply:
I don't think the layout is important.  This is a index of heap buffer for each group.  The array is indexed on the group id.... maybe stated more cleanly

Reply:
I wrote it this way because the rank and dense rank versions look similar but have more complex layouts -- and was intending to show the symmetry this way. Any suggestions on how to connect those that way?